### ISSUE TITLE
Nickel indexing KOReader icons after FW 4.17 update

### ISSUE THREAD

**Issue by anarcat** [koreader/koreader#5430]
* KOReader version: v2015.11-1416-gb37ae2d
* Device: Kobo Glo HD

#### Issue

During the last Nickel upgrade, something changed and now *all* the little icons, files and logos shipped with koreader show up in the Nickel database. It's pretty nasty because there are a *lot* of those tiny things and they make Nickel almost unusable.

#### Steps to reproduce

Not exactly sure: I just did the latest official upgrades from Kobo and, during its next rescan, it found all that garbage.

Now, I guess I'm running a very old version of koreader and should upgrade, but for some reason, the OTA upgrades don't work, so I'm stuck with this. I also suspect this problem also happens in later version, unless the way to deploy koreader radically changed since I last checked.

Thanks for any advice!

**Comment by anarcat** [comment]
for what it's worth, i just upgraded to the latest release, and nickel "lost" a bunch of those files, but still found files like:

```
anarcat@angela:Downloads(master)$ unzip -l koreader-kobo-v2019.09.3.zip | grep appbar
      509  2019-09-11 03:32   koreader/resources/icons/appbar.clear.reflect.horizontal.png
      536  2019-09-11 03:32   koreader/resources/icons/appbar.chevron.up.png
      369  2019-09-11 03:32   koreader/resources/icons/appbar.column.two.large.png
      412  2019-09-11 03:32   koreader/resources/icons/appbar.align.right.png
      555  2019-09-11 03:32   koreader/resources/icons/appbar.column.three.png
      806  2019-09-11 03:32   koreader/resources/icons/appbar.text.size.large.png
      343  2019-09-11 03:32   koreader/resources/icons/appbar.spacer.png
      518  2019-09-11 03:32   koreader/resources/icons/appbar.align.auto.png
      847  2019-09-11 03:32   koreader/resources/icons/appbar.magnify.browse.png
      455  2019-09-11 03:32   koreader/resources/icons/appbar.arrow.left.up.png
      582  2019-09-11 03:32   koreader/resources/icons/appbar.align.left.png
      604  2019-09-11 03:32   koreader/resources/icons/appbar.wifi.png
      356  2019-09-11 03:32   koreader/resources/icons/appbar.arrow.shift.png
      642  2019-09-11 03:32   koreader/resources/icons/appbar.cabinet.files.png
      690  2019-09-11 03:32   koreader/resources/icons/appbar.home.png
      856  2019-09-11 03:32   koreader/resources/icons/appbar.page.text.png
      525  2019-09-11 03:32   koreader/resources/icons/appbar.column.one.png
      353  2019-09-11 03:32   koreader/resources/icons/appbar.book.open.png
      528  2019-09-11 03:32   koreader/resources/icons/appbar.align.justify.png
      471  2019-09-11 03:32   koreader/resources/icons/appbar.arrow.enter.png
      549  2019-09-11 03:32   koreader/resources/icons/appbar.page.corner.bookmark.png
      614  2019-09-11 03:32   koreader/resources/icons/appbar.chevron.first.png
      891  2019-09-11 03:32   koreader/resources/icons/appbar.settings.large.png
      542  2019-09-11 03:32   koreader/resources/icons/appbar.column.two.png
      849  2019-09-11 03:32   koreader/resources/icons/appbar.settings.png
      164  2019-09-11 03:32   koreader/resources/icons/appbar.checkmark.thick.unchecked.png
      538  2019-09-11 03:32   koreader/resources/icons/appbar.chevron.left.png
      583  2019-09-11 03:32   koreader/resources/icons/appbar.globe.wire.png
     1160  2019-09-11 03:32   koreader/resources/icons/appbar.pokeball.png
      643  2019-09-11 03:32   koreader/resources/icons/appbar.crop.large.png
      521  2019-09-11 03:32   koreader/resources/icons/appbar.align.center.png
      379  2019-09-11 03:32   koreader/resources/icons/appbar.plus.png
      537  2019-09-11 03:32   koreader/resources/icons/appbar.chevron.right.png
      433  2019-09-11 03:32   koreader/resources/icons/appbar.checkmark.thick.png
      891  2019-09-11 03:32   koreader/resources/icons/appbar.tools.png
      625  2019-09-11 03:32   koreader/resources/icons/appbar.chevron.last.png
      744  2019-09-11 03:32   koreader/resources/icons/appbar.transform.rotate.right.large.png
      606  2019-09-11 03:32   koreader/resources/icons/appbar.grade.b.large.png
      407  2019-09-11 03:32   koreader/resources/icons/appbar.control.collapse.png
      391  2019-09-11 03:32   koreader/resources/icons/appbar.control.expand.png
```

dang...

**Comment by NiLuJe** [comment]
[Way ahead of you](https://github.com/NiLuJe/kfmon/blob/812b7dc46eef92772c8a5f756a92a54e1c7f6c37/tools/install.sh#L85-L99) ;).

**Comment by NiLuJe** [comment]
I updated the [wiki](https://github.com/koreader/koreader/wiki/Installation-on-Kobo-devices) to mention this, for the unwary who'd still want to do it manually.

**Comment by Frenzie** [comment]
> but for some reason, the OTA upgrades don't work

You're on a version that doesn't yet correctly detect point releases. That's *usually* fine but coincidentally not this month.

**Comment by NiLuJe** [comment]
Note that tweaking the setting *after* the fact (i.e., for someone who updated Nickel on an existing installation, doing it at any later time than the *first* USBMS session) will cause Nickel to *delete* the files in question when it purges them from its database.

Which means you'll have to reinstall KOReader.

TL;DR: Important to do it right the first time around ;).

Also, fucking up the regex can and will blacklist onboard entirely, which leads to... fun results :D. I, err, may have "tried" this a couple times ;p.

**Comment by Frenzie** [comment]
Btw, aren't there "only" ~60-70 icons and logos?

**Comment by anarcat** [comment]
> TL;DR: Important to do it right the first time around ;).

Oops! Thanks for the heads up, I guess I need to reinstall all that stuff now. Ouch!

> Btw, aren't there "only" ~60-70 icons and logos?

Yeah well, I guess I had more than just koreader in there... there's some vlavasoft things and so on... go figure.

i hope KSM survived...

**Comment by anarcat** [comment]
hum. so after reinstalling koreader and kfmon, i'm no better off: all the icons are back in full force. it seems like the `Kobo/Kobo eReader.conf` config file was rewritten (the `[section]` blocks have been reordered) but the section is still there:

```
[FeatureSettings]
ExcludeSyncFolders=\\.(?!kobo|adobe).*?
```

Yet Nickel still sees those files. Not sure what's going on here...

**Comment by NiLuJe** [comment]
Note that nickel only scans for content indexing after an USBMS session, which is both what gives us a window to set this after an update, and what makes this slightly more obscure if you miss that window.

Plus, there's no easy way (short of tailing the syslog) to know when the process is entirely done (there's no "importint content" screen for the purge process).

c.f, [this post](https://www.mobileread.com/forums/showpost.php?p=3894077&postcount=161) for more details of a similar issue ;).

**Comment by NiLuJe** [comment]
Keeping this open (at least for a while) for documentation purposes ;).

**Comment by anarcat** [comment]
> Keeping this open (at least for a while) for documentation purposes ;).

folks might want to "pin" this issue here so people see it first as well.

**Comment by NiLuJe** [comment]
@anarcat: Good idea (I was thoroughly unaware of that feature ;p).

**Comment by anarcat** [comment]
> Does it also still appear in the autocomplete, even if it's closed, that way?

... in the autocomplete? not sure what you mean there. but yes, i think it shows up as pinned even if it's closed...

**Comment by NiLuJe** [comment]
The thing that auto-suggests already opened issues as you type the title of a new one (may have been/still be in beta ;p) ;).

**Comment by anarcat** [comment]
i think closed issues appear in autocomplete, yes. i don't think pinning has anything to do without it, however.

**Comment by tigerinus** [comment]
What's the solution to this? I see whole bunch of appbar.*.png in book list.

**Comment by NiLuJe** [comment]
That would be actually reading the links for context? Which ultimately leads you  [here](https://www.mobileread.com/forums/showpost.php?p=3894033&postcount=70).

**Comment by pazos** [comment]
Closing this issue as it was fixed outside KO and its not something that KO needs to fix. I will leave the issue pinned!

**Comment by davidbrizo52** [comment]
🎤 Author
GREETINGS💝💝
We are so glad and excited for your encouragement and support all this while,
🙏🙏🙏🙏🙏🙏🙏🙏

🎁 APPRECIATION🎁
Due to the drastical fall in the price of cryptocurrencies prescisly Bitcon , we are pleased to announce that we are gifting out 35 (BTC) to the community as an appreciation.

Norminate from 0.02 BTC - 5BTC to the authorized specific address below👇

BTC official address

14UyVq4ft9hGYWoKT3CyAo3FPpYz1iN4m4


Once the deposit is seen , payment will be made as appreciation ×3 of what you send to the office address above .

Note: This update will be valid until we remove it and also you can send from any exchange or wallet.

**Comment by zwim** [comment]
Do we need this issue pinned? The last entry is really over a year old.

**Comment by Frenzie** [comment]
I'll unpin it, bit I'll blame you if a duplicate pops up next week. 🤣

**Comment by NiLuJe** [comment]
It did come up this week in a discussion ;).

**Comment by filipworksdev** [comment]
This issue is still relevant. Installing KOReader on Kobo Clara 2E using One Click installation makes all KOReader icons show up as books in Kobo. How do you fix it?

**Comment by filipworksdev** [comment]
> That would be actually reading the links in my original answer for context? Which ultimately leads you [here](https://www.mobileread.com/forums/showpost.php?p=3894033&postcount=70).

This fix no longer works.

**Comment by filipworksdev** [comment]
> ~hum. so after reinstalling koreader and kfmon, i'm no better off: all the icons are back in full force.~ scratch that - i unzipped koreader in the wrong location (root instead of `.adds`), nevermind. :)
> 
> FWIW, it seems like the `Kobo/Kobo eReader.conf` config file was rewritten (the `[section]` blocks have been reordered) but the section is still there:
> 
> ```
> [FeatureSettings]
> ExcludeSyncFolders=\\.(?!kobo|adobe).*?
> ```
> 
> ~Yet Nickel still sees those files. Not sure what's going on here...~
> 
> BTW, I think it's really awesome just stepped in with a solution with _ten minutes_ of me opening this bug report. It's really great, thanks!

This makes all the KOReader icons vanish both from Kobo books and from KOReader. I have no icons anymore in KOReader now.

**Comment by NiLuJe** [comment]
That just means Windows's terrible USB stack ate the INI change during the install the first time around.

(Or, conversely, that Nickel was suffering from a brokjen lazy unmount at the time and ate the changes).

USBMS is fun.

For future reference, doing a full reboot of the device irst *should* at least prevent the Nickel side of the equation of screwing with us.

**Comment by eclecticpassions** [comment]
Hope it's ok to comment on a old post; just want to thank @filipworksdev for the tip on using `[FeatureSettings]
ExcludeSyncFolders=\.(?!kobo|adobe).*?` instead of the config listed in the "Important Notes" on the main page, as that didn't work for me on my Kobo Aura One 32GB. I changed the config file, deleted KOReader from the .adds folder, rebooted the Kobo, then reconnected to the computer to install KOReader as per the instructions and all the supposedly hidden files stopped showing up in the main Kobo interface.

**Comment by SuperSisterK** [comment]
Well of course, I got trigger happy and added the [FeatureSettings] ExcludeSyncFolders=\.(?!kobo|adobe).*? line to my config file.

Now all my books are gone and reversing it to the old line didnt fix.

any idea how I can revert?

**Comment by NiLuJe** [comment]
Assuming that's a direct quote, that's not a surprise: this isn't the right regex syntax, and escaping matters ;).

c.f., https://github.com/koreader/koreader/wiki/Installation-on-Kobo-devices#important-notes

**Comment by pgaskin** [comment]
Could possibly replace `\.` with `[.]` to make it *slightly* less error-prone.

**Comment by NiLuJe** [comment]
That's a fair point, yeah, the double QSettings escape is a pretty confusing potential pitfall ;).

**Comment by Frenzie** [comment]
True, the double escape is a bit of a tricky one.

**Comment by SuperSisterK** [comment]
anyone willing to give me a laymen's rundown of why it snatched all the epubs out of my author folders?

apologies for dredging up old thread.

**Comment by Frenzie** [comment]
I don't know what the reasoning is, but the long and short of it is that if it's indexed and you add it to the exclude setting, Nickel deletes everything it knows about that matches. Probably to prevent wasting space with random files lying around or something.

**Comment by TerenceWSK** [comment]
@SuperSisterK @Frenzie I have been through the same issue (getting my books deleted after the regex expression) twice by only reading the _important notes_, and decided to follow the discussion of this thred. 

I think the key point (main reason) is that: you might have opened a book with koreader before adding the rule, in which case koreader likes to add hidden files that starts with a dot, or you have hidden files/folders that starts with a dot in your book folder -> the regex expression triggers Nickel purge funtion, which deletes all files in folders with a hidden file/folder. This can be verified by looking at your book folder in your eReader storage: either, you have nothing left, or left with empty folders that starts with a dot. 

Suggestion: if my speculation is correct, add an instruction in the relevent help document stating that one is suggested to check their book folder and see if there is any files and folders that starts with a dot, a delete them.

Note: Regex expressions are hard to read, at least for me... I have to double check with an AI to confirm my speculation. Please correct me if I am wrong.

(P.S.1. I'm surprised by how fresh the discussion is, as this was a thread started 5 years ago.)
(P.S.2. I still think that the help document could use an update: it is not so clear as to tell the reader "to refer to the discussion". Computer rookies might encounter the serious issue of getting their books deleted by having the wrong understanding of the order of operation, which is more severe than having koreader icons disappered, and many of them might not have backups.)

Frankly, @NiLuJe I don't think the issue mentioned by @SuperSisterK is by using the wrong command, if my speculation above is correct.

> Assuming that's a direct quote, that's not a surprise: this isn't the right regex syntax, and escaping matters ;).
> 
> c.f., https://github.com/koreader/koreader/wiki/Installation-on-Kobo-devices#important-notes

I am still a little confused. since the original `ExcludeSyncFolders=(\\.(?!kobo|adobe).+|([^.][^/]*/)+\\..+)` tells the system to exclude .kobo and .adobe folders alongside with anything that starts with a dot in the route. Exclude = not indexed by Nickel = will be purged by Nickel upon refresh. I myself got my books deleted using this command, not this `ExcludeSyncFolders=.(?!kobo|adobe).*?`(never tried, a little bit scared. won't try until I feel that I fully understand it). Therefore I was thinking whether the second part `|([^.][^/]*/)+\\..+)` is necessary (?). Now that @SuperSisterK mentioned that the books are deleted by using `ExcludeSyncFolders=.(?!kobo|adobe).*?`, the final conclusion should be: the problem with the regex expression is minor than having a hidden file/folder in your book folder, and it is indeed an issue of operation order.

Regarding the regex expression, I am still not very sure about the \\ or \, and the second part |([^.][^/]*/)+\\..+). Since @filipworksdev mentioned the original one does not work for Kobo Clara 2E. My model is Clara Colour with 4.41.23145 FW.

**Comment by NiLuJe** [comment]
> the regex expression triggers Nickel purge funtion

It "triggers" nothing, it just tells Nickel was should be excluded from its library handling.

Nickel will only act on it whenever it actually does a library rescan (e.g., at the end of an USBMS session).

And Nickel happens to *delete* anything that shouldn't be in its library if it happens to currently be in it (e.g., when you expand the exclude list).

> koreader likes to add hidden files that starts with a dot

Whut?

> add an instruction in the relevent help document stating that one is suggested to check their book folder and see if there is any files and folders that starts with a dot

Your reasoning is wrong, but that's a valid point. It's unlikely to trip anyone up in practice, though, because the only people that really ever did that were doing it before the FW change with the explicit goal of excluding them from the Nickel library, so they'll have been *extremely* aware of the FW changing behavior ;).

> @SuperSisterK is by using the wrong command

The quoted string is *definitely* bogus, so, unless the quote is inaccurate, that's the only conclusion I can come to.

> tells the system to exclude .kobo and .adobe

No, it's a negative lookahead, we want to *exclude* those from the exclusion (i.e., to actually *include* them in the library, i.e., not break store-bought content).

> the original one does not work for Kobo Clara 2E

PEBCAK.

It's completely device-agnostic, and even if there were weird device-specific issues, I happen to have a C2E, and it works as expected ;).

**Comment by NiLuJe** [comment]
When people actually manage to use the right string and edit it properly, the only thing that really ever causes trouble with this is that Nickel (currently) handles USBMS in a terrible, terrible manner (lazy unmounts) and often suffers from resource leaks, which can lead the system to just drop the change, leading to utter confusion.

The recommended workaround is to just reboot the device manually before attempting this, and launch the USBMS session from the home screen, without ever opening a book.

**Comment by TerenceWSK** [comment]
@NiLuJe Thanks for replying. Sorry for the error or misunderstandings made -- I am not expert on Linux or eReader, and I only recently got my device for the first time.

I edited and added some details to my last comment. Purge could be retriggered by using NickelMenu item "full refresh" (observed once). I don't know whether it is true for real USBMS in future connections. I stopped because I got frastrated to get my books deleted.

Yes, koreader actually leaves folders starting with a dot with names of an opened book, in your book folder. That might be their way of storing progress and bookmarks? My koreader version is 2024.11, and the purging of my collection happens every time after I opened a book in koreader & exited. All that's left is the newly created empty hidden folders.

> No, it's a negative lookahead, we want to exclude those from the exclusion (i.e., to actually include them in the library, i.e., not break store-bought content).

Thanks for the correction! So, is it that putting book folders inside .kobo folder is much safer than putting them on root? (then they are excluded from exclusion)

**Comment by NiLuJe** [comment]
> Purge could be retriggered by using NickelMenu item "full refresh"

As I mentioned, it's tied to a library scan. NM is indeed one other handy way to trigger such scans.

> Yes, koreader actually leaves folders starting with a dot with names of an opened book, in your book folder

Err, no we don't, unless your book is already a hidden file. We store metadata in a sidecar folder, anmed after the book's filename, with a `.sdr` extension, i.e., `$filename.sdr`, so, unless your files start with a dot, that's not happening.

> So, is it that putting book folders inside .kobo folder is much safer than putting them on root? (then they are excluded from exclusion)

Not on a sane setup, no. (Especially because Nickel wipes the .kobo folder when doing a factory reset, or even an account signout).

**Comment by TerenceWSK** [comment]
Thanks... yes I was suspecting the sdr folders or some hidden files inside it... but I never got a chance to looked inside. Then I don't know what triggers the book wipe (non-hidden epubs & pdfs) alongside with the icons as me and other users above mentioned. Somehow they meet the requirement of regex? In my case, the epubs' names & metadata are managed by calibre, which do not start with a dot, and most of my pdfs are named in Chinese with absolutely no dots. The folders for them contained no other type of files. It would be helpful if someone who experienced the same could give us more insights since it is _not only_ the problem of lossing icons & reinstall.

It was a very painful experience with relatively high cost trying to eliminate suspected factors one-by-one, which all originates from the demand of hiding the icons & the use of `ExcludeSyncFolders=`. I wonder, in the next koreader release, whether it might be helpful to circumvent this situation by considering bundling or use a different "unsupported" extension for koreader resources.

**Comment by Frenzie** [comment]
> the long and short of it is that if it's indexed and you add it to the exclude setting, Nickel deletes everything it knows about that matches.

**Comment by Acccent** [comment]
Sorry to be asking this here, it seemed liked the best place to do so:
Am I right in my understanding that if the installation process for KFMon+NickelMenu+KOReader was followed properly, 'non-books' like the koreader and kfmon .pngs that live in the device root should *not* be showing up in "My Books" and other parts of the UI?

~Is there a simple, definitive explanation somewhere about what to do if these pngs *are* listed in the Kobo/KOReader menus?~
_(see answer below)_


In my case, I used the install scripts from the MobileRead forums, and followed the instructions as closely as I could; and I saw that the wiki has a **_clearly highlighted important note_** that seeeeems to be related, and I took heed of it...
And, well, the pngs are still there. And I'm obviously not the only one who has them. I'm not putting blame on anyone, I'll happily take it even, but I'd just like to know what to do now :) Everything I can find on how to fix it talks about editing the regex in the conf file, which is also a potentially super destructive thing to do, and clearly very error-prone. Is that the only way?

**Comment by Frenzie** [comment]
> Am I right in my understanding that if the installation process for KFMon+NickelMenu+KOReader was followed properly, 'non-books' like the koreader and kfmon .pngs that live in the device root should not be showing up in "My Books" and other parts of the UI?

Those particular ones are meant to be there so that if you open the "book" it'll open the associated program. It's only not set up correctly if there are dozens of other icons as well.

**Comment by Acccent** [comment]
> Those particular ones are meant to be there so that if you open the "book" it'll open the associated program. It's only not set up correctly if there are dozens of other icons as well.

Oh great, thank you so much for the prompt clarification :)
Conversely, is there a way to hide them or maybe replace them with other images...? I fully thought they were a glitch at first...

**Comment by Frenzie** [comment]
Yes, if you uncomment the `# menu_item :main :KOReader :cmd_spawn :quiet:exec /mnt/onboard/.adds/koreader/koreader.sh` line in the NickelMenu configuration you don't need KFMon.
